spec/prog/ai/inference_endpoint_replica_nexus_spec.rb

```diff
@@ -0,0 +1,215 @@
+      expect(nx).to receive(:when_destroy_set?).and_yield
+      expect(nx.strand).to receive(:label).and_return("destroy")
+      expect { nx.before_run }.not_to hop("destroy")
+    end
+
+    it "pops additional operations from stack" do
+      expect(nx).to receive(:when_destroy_set?).and_yield
+      expect(nx.strand).to receive(:label).and_return("destroy")
+      expect(nx.strand.stack).to receive(:count).and_return(2)
+      expect { nx.before_run }.to exit({"msg" => "operation is cancelled due to the destruction of the inference endpoint replica"})
+    end
+  end
+
+  describe "#start" do
+    it "naps if vm not ready" do
+      expect(replica.vm).to receive(:strand).and_return(instance_double(Strand, label: "prep"))
+      expect { nx.start }.to nap(5)
+    end
+
+    it "update sshable host and hops" do
+      expect(replica.vm).to receive(:strand).and_return(instance_double(Strand, label: "wait"))
+      expect { nx.start }.to hop("bootstrap_rhizome")
+    end
+  end
+
+  describe "#bootstrap_rhizome" do
+    it "buds a bootstrap rhizome process" do
+      expect(nx).to receive(:bud).with(Prog::BootstrapRhizome, {"target_folder" => "inference_endpoint", "subject_id" => replica.vm.id, "user" => "ubi"})
+      expect { nx.bootstrap_rhizome }.to hop("wait_bootstrap_rhizome")
+    end
+  end
+
+  describe "#wait_bootstrap_rhizome" do
+    before { expect(nx).to receive(:reap) }
+
+    it "hops to setup if there are no sub-programs running" do
+      expect(nx).to receive(:leaf?).and_return true
+
+      expect { nx.wait_bootstrap_rhizome }.to hop("download_lb_cert")
+    end
+
+    it "donates if there are sub-programs running" do
+      expect(nx).to receive(:leaf?).and_return false
+      expect(nx).to receive(:donate).and_call_original
+
+      expect { nx.wait_bootstrap_rhizome }.to nap(1)
+    end
+  end
+
+  describe "#download_lb_cert" do
+    it "downloads lb cert and hops to setup" do
+      expect(sshable).to receive(:cmd).with("sudo inference_endpoint/bin/download-lb-cert")
+      expect { nx.download_lb_cert }.to hop("setup")
+    end
+  end
+
+  describe "#setup" do
+    it "triggers setup if setup command is not sent yet or failed" do
+      expect(sshable).to receive(:cmd).with("common/bin/daemonizer 'sudo inference_endpoint/bin/setup-replica' setup", {stdin: "{\"inference_engine\":\"vllm\",\"inference_engine_params\":\"--some-params\",\"model\":\"llama\",\"replica_ubid\":\"theubid\",\"ssl_crt_path\":\"/ie/workdir/ssl/ubi_cert.pem\",\"ssl_key_path\":\"/ie/workdir/ssl/ubi_key.pem\",\"gateway_port\":8443}"}).twice
+      expect(inference_endpoint).to receive(:engine).and_return("vllm").twice
+      expect(inference_endpoint).to receive(:engine_params).and_return("--some-params").twice
+      expect(inference_endpoint).to receive(:model_name).and_return("llama").twice
+      expect(inference_endpoint).to receive(:load_balancer).and_return(instance_double(LoadBalancer, id: "lb-id", dst_port: 8443)).twice
```

Comment:
fyi, you may be able to condense this with instance_double (which doesn't care about `twice`, which I tend to view as overfitting unless there's something very special about the logic under test) or `at_least(:once)` (which I really wish was less wordy, but handles the case where one is indifferent to the number of times it is called)